### PR TITLE
Refactor CMA-ES sampler with search space transform

### PR DESCRIPTION
**Pull request by Crissman**
## Motivation

Reduce the code required for CMA-ES sampler by leveraging the search space transform code.

## Description of the changes

Use the search space transform code to simplify the CMA-ES sampler.

### CODE REVIEW

**Comment by hvy** [comment]
Nice to see more usages of the search space transform. Is this ready for review? It might otherwise be better to make it a draft PR or prefix the title with something like "[WIP]" (saying this since it seems like you're still working on it).

**Comment by Crissman** [comment]
Confirmed this PR performed comparable to existing CMA-ES sampler with 100 trials.

Ready for review.

**Comment by hvy** [comment]
As a follow up to this PR, we could start allowing categoricals with this sampler. We just need to add a single line to `infer_relative_search_space` including `CategoricalDistribution` since the transformer already takes care of the one-hot encodings and necessary conversions.

**Comment by hvy** [comment]
Let me ping @c-bata as he's the author of the sampler. He might have some inputs or feedbacks on the changes (if you have time).

File: optuna/samplers/_cmaes.py [hunk 1]
```diff
@@ -352,38 +358,35 @@ def _restore_optimizer(
...
         if randomize_start_point:
             # `_initialize_x0_randomly ` returns internal representations.
-            x0 = _initialize_x0_randomly(self._cma_rng, search_space)
-            mean = np.array([x0[k] for k in ordered_keys], dtype=float)
+            mean = np.array([b[0] + (b[1] - b[0]) * self._cma_rng.rand() for b in bds])
         elif self._x0 is None:
             # `_initialize_x0` returns internal representations.
-            x0 = _initialize_x0(search_space)
-            mean = np.array([x0[k] for k in ordered_keys], dtype=float)
+            mean = np.array([b[0] + (b[1] - b[0]) / 2 for b in bds])
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:373]: How about calculating a mean vector like the following to remove list comprehensions?

```python
lower_bounds = trans.bounds[:, 0]
upper_bounds = trans.bounds[:, 1]
n_dimension = len(trans.bounds)

if randomize_start_point:
    mean = lower_bounds + (upper_bounds - lower_bounds) * self._cma_rng.rand(n_dimension)
elif self._x0 is None:
    mean = lower_bounds + (upper_bounds - lower_bounds) / 2
```

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request!

The initial impression is that the changes are almost perfect 👍 
I put a review comment.

File: optuna/samplers/_cmaes.py (hunk 1 above)
Comment (Crissman) [reply optuna/samplers/_cmaes.py]: Done. Thanks!

File: optuna/samplers/_cmaes.py [hunk 2]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
-            sigma0 = _initialize_sigma0(search_space)
+            sigma0 = min([(b[1] - b[0]) / 6 for b in bds])
         else:
             sigma0 = self._sigma0
 
         # Avoid ZeroDivisionError in cmaes.
         sigma0 = max(sigma0, _EPS)
-        bounds = _get_search_space_bound(ordered_keys, search_space)
-        n_dimension = len(ordered_keys)
+        n_dimension = len(bds)
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:390]: `n_dimension` is already defined at L369.

```suggestion
```

File: optuna/samplers/_cmaes.py [hunk 3]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
     def _init_optimizer(
         self,
-        search_space: Dict[str, BaseDistribution],
-        ordered_keys: List[str],
+        trans: _SearchSpaceTransform,
         population_size: Optional[int] = None,
         randomize_start_point: bool = False,
     ) -> CMA:
+
+        bds = trans.bounds
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:366]:
```suggestion
```

File: optuna/samplers/_cmaes.py [hunk 4]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
+        lower_bounds = trans.bounds[:, 0]
+        upper_bounds = trans.bounds[:, 1]
+        n_dimension = len(trans.bounds)
+
         if randomize_start_point:
             # `_initialize_x0_randomly ` returns internal representations.
-            x0 = _initialize_x0_randomly(self._cma_rng, search_space)
-            mean = np.array([x0[k] for k in ordered_keys], dtype=float)
+            # mean = np.array([b[0] + (b[1] - b[0]) * self._cma_rng.rand() for b in bds])
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:373]: It doesn't need to keep this line.

```suggestion
```

File: optuna/samplers/_cmaes.py [hunk 5]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
+            mean = lower_bounds + (upper_bounds - lower_bounds) * self._cma_rng.rand(n_dimension)
         elif self._x0 is None:
             # `_initialize_x0` returns internal representations.
-            x0 = _initialize_x0(search_space)
-            mean = np.array([x0[k] for k in ordered_keys], dtype=float)
+            # mean = np.array([b[0] + (b[1] - b[0]) / 2 for b in bds])
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:377]: ditto.

```suggestion
```

File: optuna/samplers/_cmaes.py [hunk 6]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
         else:
             # `self._x0` is external representations.
-            mean = np.array(
-                [_to_cma_param(search_space[k], self._x0[k]) for k in ordered_keys], dtype=float
-            )
+            mean = trans.transform(self._x0)
 
         if self._sigma0 is None:
-            sigma0 = _initialize_sigma0(search_space)
+            sigma0 = min([(b[1] - b[0]) / 6 for b in bds])
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:384]: You can calculate sigma without list comprehensions.

```suggestion
            sigma0 = np.min((upper_bounds - lower_bounds) / 6)
```

File: optuna/samplers/_cmaes.py [hunk 7]
```diff
@@ -352,38 +358,40 @@ def _restore_optimizer(
...
         # Avoid ZeroDivisionError in cmaes.
         sigma0 = max(sigma0, _EPS)
-        bounds = _get_search_space_bound(ordered_keys, search_space)
-        n_dimension = len(ordered_keys)
+        n_dimension = len(bds)
         return CMA(
             mean=mean,
             sigma=sigma0,
-            bounds=bounds,
+            bounds=bds,
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:394]:
```suggestion
            bounds=trans.bounds,
```

**Review by c-bata — COMMENTED** [review]
Thank you for your swift update. I have some suggestions for refactoring. Could you check them?

**Comment by Crissman** [comment]
Thanks, @c-bata . That's cleaner. 🙇‍♂️

**Review by c-bata — APPROVED** [review]
LGTM! Thank you for your pull request.

File: optuna/samplers/_cmaes.py [hunk 8]
```diff
@@ -352,38 +358,36 @@ def _restore_optimizer(
...
-        search_space: Dict[str, BaseDistribution],
-        ordered_keys: List[str],
+        trans: _SearchSpaceTransform,
         population_size: Optional[int] = None,
         randomize_start_point: bool = False,
     ) -> CMA:
+
+        lower_bounds = trans.bounds[:, 0]
+        upper_bounds = trans.bounds[:, 1]
+        n_dimension = len(trans.bounds)
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:368]: I'm sorry. I found a bug. The shape of `trans.bounds` is `(2, n_dimensions)`. So we need to calculate n_dimension like the following.

```suggestion
        n_dimension = len(lower_bounds)
```

File: optuna/samplers/_cmaes.py [hunk 9]
```diff
@@ -352,38 +358,36 @@ def _restore_optimizer(
...
+
         if randomize_start_point:
             # `_initialize_x0_randomly ` returns internal representations.
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:371]: I notice that we can remove this comment because `_initialize_x0_randomly` is removed. L374 is also.

**Review by c-bata — APPROVED** [review]

**Review by c-bata — APPROVED** [review]
Thank you!

**Review by HideakiImamura — APPROVED** [review]
Thanks for the amazing simplification! Looks perfect!

**Comment by HideakiImamura** [comment]
Let me merge this PR since we have two approvals.

**Pull request merged**